### DISCUSSION TITLE
Is there a plan to support Qt (QWindow)?

### DISCUSSION THREAD

**Discussion by zhangfq-chemistry** [MethanePowered/MethaneKit#144] [Q&A] [answered]

**Comment by egorodet** [comment, marked as answer]
@zhangfq-chemistry Hello, currently I don't have short-term plans to support Qt. MethaneKit aims to provide minimal platform support required for graphics rendering with as simple as possible implementation, based on native APIs without using any 3rd-party libraries requiring dynamic linking or provided under commercial licenses. I may consider supporting Qt in future or by external contributors, since it should not be difficult to do: all is needed is to pass window handles to RHI and to translate input events. Could you please describe your particular use case and reason why you want to use this library with Qt?

**Reply by zhangfq-chemistry** [reply]
I use Qt (non-commercial)  for years due to its powerful UI/widgets. What I need is just to replace the simple qt-opengl-widget  with  methaneKit. Thanks a lot for your wonderful project!

**Reply by egorodet** [reply]
Latest versions of Qt offer their own graphics abstraction layer on top of modern graphic APIs, also called Qt RHI. This RHI was previously used internally for Qt Quick rendering and recently became publicly available. Also Qt provides several examples of using their Qt RHI inside classic Qt with Widgets, take a look at this examples:
 - [RHI Window Example](https://doc.qt.io/qt-6/qtgui-rhiwindow-example.html)
 - [Simple RHI Widget Example](https://doc.qt.io/qt-6/qtwidgets-rhi-simplerhiwidget-example.html)

So, Qt already provides similar modern graphics abstraction layer perfectly integrated in their library. Is that what you was looking for?

**Comment by zhangfq-chemistry** [comment]
I use RHI/Qt for two years, and it only provide basic functions. In contrast, the MethaneKit is more powerful. VulkanSceneGraph also provide vsgQt to support Qt, and I find  it is over-encapsulation.

**Reply by egorodet** [reply]
Ok, thank you for clarification. I will consider Qt integration for the future. Feature request #145 was created.

**Comment by zhangfq-chemistry** [comment]
Thanks a lot!

**Reply by zhangfq-chemistry** [reply]
One solution from Nicholas Angle, but I do not know how to obtain vulkanInstance from MethaneKit.
pseudo-code overview of all the steps:
{
QVulkanInstance instance;
instance.setVkInstance(vulkanInstanceFromRenderer);
instance.create();
QWindow* window = new QWindow;
window->setSurfaceType(QSurface::VulkanSurface);
window->setVulkanInstance(instance);
QWidget* wrapper = QWidget::createWindowContainer(window, parentWidget);
VkSurfaceKHR surface = QVulkanInstance::surfaceForWindow(window);
void UIClass::resizeEvent(QResizeEvent* event) {
    window->setMinimumSize(ui->widget->size());
}

 https://www.niangames.com/articles/qt-vulkan-renderer#:~:text=By%20default%20QVulkanInstance.create%20%28%29%20will%20create%20a%20new,use%20the%20same%20VkInstance%20as%20the%20Vulkan%20renderer.
